File: examples/async_await.go
package main

import (
	"errors"
	"fmt"
	"os"
	"time"
)

func hello(name string) chan interface{} {
	ch := make(chan interface{}, 1)
	go func() {
		time.Sleep(1 * time.Second)
		if name == "fail" {
			ch <- errors.New("failed")
		} else {
			ch <- "hello " + name
		}
	}()

	return ch
}

func main() {
	result := <-hello("bob")
	switch v := result.(type) {
	case string:
		fmt.Println(v)
	case error:
		fmt.Fprintln(os.Stderr, v)
	}

	result = <-hello("fail")
	switch v := result.(type) {
	case string:
		fmt.Println(v)
	case error:
		fmt.Fprintln(os.Stderr, v)
	}
}

File: examples/async_await.c
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static const char *hello(const char *name, bool *failed) {
    *failed = strcmp(name, "fail") == 0;
    return *failed ? "failed" : "hello bob";
}
int main(void) {
    bool failed = false;
    const char *result = hello("bob", &failed);
    if (failed) fprintf(stderr, "%s\n", result); else printf("%s\n", result);
    result = hello("fail", &failed);
    if (failed) fprintf(stderr, "%s\n", result); else printf("%s\n", result);
    return 0;
}
